docs: add CONTRIBUTING, DEVELOPING and VSCODE

bajtos opened the pull request:
Start a guide for contributors in a place that's [recognized by GitHub](https://help.github.com/articles/setting-guidelines-for-repository-contributors/) - add a short CONTRIBUTING.md file serving as a collection of links pointing to other documentation.

Start a guide for developers in DEVELOPING.md.

Describe VS Code setup in VSCODE.md, document steps needed to verify
integration with tsc and tslint.

Add docs/apidocs.html as an index file pointing to apidoc files of individual monorepo packages.

This work is loosely related to https://github.com/strongloop/loopback-next/issues/950 and #574. Even though it is not in MVP scope, I think it's very important for us internally, to avoid regressions like the one introduced by https://github.com/strongloop/loopback-next/pull/964

I am intentionally **not wrapping** lines at 80 characters, because this content will be rendered by GitHub and such line-breaks would be preserved in the rendered output.

The text is inspired by the following guides:
  - Node.js [pull-requests.md](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md) and [CONTRIBUTING.md](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md)
 - AngularJS [DEVELOPERS.md](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md)

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->


## Checklist

- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] Related API Documentation was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `packages/example-*` were updated

bajtos commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,117 @@
+7. Open "Tasks" OUTPUT window and verify that compilation error was parsed by VSCode.
+
+8. Verify that compilation errors are correctly associated with the problematic source code line.
+   _(2018-02-82: this is does not work now, `tsc` is reporting paths relative to individual package directories.)_
```
FWIW: this used to work before, I suspect https://github.com/strongloop/loopback-next/pull/682 broke it.

The trick is that we need to run `tsc` from the root monorepo directory. 

https://github.com/strongloop/loopback-next/pull/682/files#diff-272a4137cf0e755785f43213ae0bae7fL14
> Under the hood, we change the current directory to loopback-next root
> and call tsc with project configuration from the original package's
> tsconfig.build.json. This way tsc error messages contain a path
> that's relative to loopback-next project root.

https://github.com/strongloop/loopback-next/pull/682/files#diff-272a4137cf0e755785f43213ae0bae7fL60

```
'-p',
// It's important to keep this path relative to rootDir
path.join(packageDir, 'tsconfig.build.json'),
```

I have some ideas how to fix this (basically introduce a new `lb-tsc` argument like `-r ../..` that will control the working directory of `tsc`), but I prefer to leave such changes out of this documentation-only pull request.

bajtos commented:
Thank you @virkt25 for a thoughtful comment. I fully agree with you
that we should not make the impression that loopback-next can be developed
only using VisualStudio Code.

I am a long time user of ViM which I still basically for everything except
loopback-next, so I understand your concerns.

As I was thinking about this a bit more, I agree with your that most of my
new content does not really belong to CONTRIBUTING.md. I am going to reorganize
it and let you know when I have new version to review.

bajtos commented:
@virkt25 I reorganized the content and add more details, PTAL again.

@strongloop/sq-lb-apex @strongloop/loopback-next This pull request is starting CONTRIBUTING, DEVELOPERS and VSCODE guides, please take a look and let me know if you have any feedback.

I'd like to keep the scope of this pull request small please. IMO, an incomplete guide is better than no guide. Let's land something meaningful quickly and then we can incrementally add more content as needed.

bajtos commented:
You can preview the new content here:

 - https://github.com/strongloop/loopback-next/blob/docs/describe-how-to-test-tsc-tslint-changes/docs/CONTRIBUTING.md
 - https://github.com/strongloop/loopback-next/blob/docs/describe-how-to-test-tsc-tslint-changes/docs/DEVELOPING.md
 - https://github.com/strongloop/loopback-next/blob/docs/describe-how-to-test-tsc-tslint-changes/docs/VSCODE.md

A screenshot of apidocs.html:

<img width="1056" alt="screen shot 2018-02-12 at 14 20 04" src="https://user-images.githubusercontent.com/1140553/36098612-dc1ac784-0fff-11e8-9287-534d396ac11f.png">

virkt25 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,41 @@
+# Contributing to LoopBack
+
+Contributions to Node.js include code, documentation, answering user questions, and advocating for all types of Node.js users. See our official documentation on loopback.io for more information common to all of our GitHub repositories:
```
Contributing to LoopBack!

virkt25 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,41 @@
+## [Contributing code](http://loopback.io/doc/en/contrib/code-contrib.html)
+
+Pull Requests are the way concrete changes are made to the code, documentation
+and tools contained in LoopBack repositories.
```
This is on a new line

dhmlau commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,41 @@
+ - [Using Jekyll](http://loopback.io/doc/en/contrib/jekyll_getting_started.html)
+ - [Authoring pages](http://loopback.io/doc/en/contrib/pages.html)
+ - [Translations](http://loopback.io/doc/en/contrib/translation.html)
+
```
I'm thinking we might need to add the commit message guidelines to here as well.  
e.g. https://loopback.io/doc/en/contrib/git-commit-messages.html
There were quite a number of doc PRs at the end of last year that failed in the commit linter.

virkt25 replied:
LoopBack Next uses different rules for commit messages and @bajtos has included details for that in `DEVELOPING.md`.

dhmlau replied:
If the doc PR is in loopback.io, I think the commit message guideline from https://loopback.io/doc/en/contrib/git-commit-messages.html is still valid.  
But yes, if it's from one of the READMEs, then it will follow the one in loopback-next.

bajtos replied:
I am going to add a link to that other page too.

```md
[Git commit messages](./DEVELOPING.md#commit-message-guidelines) (Note that other repositories like [strongloop/loopback.io](https://github.com/strongloop/loopback.io) use a different [convention](https://loopback.io/doc/en/contrib/git-commit-messages.html).)
```

bajtos replied:
On the second thought, I am going to revert the change mentioned in my previous comment and move the link to loopback.io commit message guidelines to the actual text in DEVELOPING.md file. A new commit will come shortly.

dhmlau commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,41 @@
+Pull Requests are the way concrete changes are made to the code, documentation
+and tools contained in LoopBack repositories.
+
+ - [Setting up development environment](./DEVELOPING.md#setting-up-development-environment)
```
Since most of the links are coming from `DEVELOPING.md`, I wonder if we could group them to a category sth like "what you need to know to develop LB".  It might be more convenient to the users to go to that link and read rather than going to each link and basically pointing to the same page.
Maybe it's just me...

richardpringle replied:
I actually like the separate links (even though they are all pointing to the same page).

bajtos replied:
Here is my point of view:

 - "what you need to know to develop LB" is already grouped in `DEVELOPING.md`. Users wishing to read everything about DEVELOPING in one go can already read that file!
 - I could shorten this list and replace all items pointing to different sections of DEVELOPING with a single link pointing to that file, but I think that's worse user experience because users cannot quickly jump from CONTRIBUTING to a relevant section in DEVELOPING.

I see two kinds of users:

 - Jane would like to learn everything about developing loopback-next in one go. She can open DEVELOPING file and read through it.
 - Joe prefers to jump to the code and read only as much as is needed. Links in CONTRIBUTING makes it easier for him to see what different topics are available and choose to read only what's relevant for him.

FWIW, Node.js uses similar approach in their [CONTRIBUTING.md](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md)

virkt25 commented on docs/DEVELOPING.md:
````diff
@@ -0,0 +1,192 @@
+we recommend to do a full reinstall of all npm dependencies:
+
+```sh
+$ npm run clean:lerna && npm run bootstrap
````
I think running just clean with bootstrap is fine. `npm run clean && npm run bootstrap`. Any reason to use `clean:lerna` here?

bajtos replied:
`clean:lerna` removes `node_modules` files in all package folders (something like `rm -rf packages/*/node_modules`).

I vaguely remember having issues with dependency setup that were resolved by removing all node_modules folders.

If this is no longer a problem, then I am happy to change this line as you suggested.

@raymondfeng could you please chime in? AFAICT, `clean:lerna` was added by you in f1e59a2e2, which says:

> This change fixes the issue that `npm run release` fails if a new package is added to the monorepo before bootstrap is done. The new package doesn't have dependencies installed yet.

Can we safely assume that `clean:lerna` script is needed only for the release process?

raymondfeng replied:
`clean:lerna` is mostly for `release` but it also be helpful in certain situations to force a fresh `bootstrap`.

richardpringle commented on docs/DEVELOPING.md:
```diff
@@ -0,0 +1,192 @@
+TypeScript too.
+
+ - [VisualStudio Code](./VSCODE.md)
+ - _Missing your favorite IDE/editor here? Pull requests are welcome!_
```
👍

bajtos replied:
On the second thought, I am going to rewrite and expand the sentence `Pull requests are welcome!`, it has built negative connotations over the years.

richardpringle commented on docs/DEVELOPING.md:
```diff
@@ -0,0 +1,192 @@
+
+#### scope
+
+The **scope** must be a list of one or more packages contained in this monorepo. Eeach scope name must match a directory name in [packages/](../packages), e.g. `core` or `context`.
```
>Eeach

woops

bajtos commented:
Thank you @virkt25 @dhmlau and @richardpringle for your comments!

I have rebase this patch on top of the latest master (50cfbcc) and addressed your comments in 3b72ced.

On the second thought, I decided to move `CONTRIBUTING.md` to the root directory of our repository (d2fbc8f), to give it a more prominent place making it easier to discover by people looking around our GitHub repo. (This is what Node.js project does too, FWIW).

LGTY now?

bajtos commented:
New preview links:
- https://github.com/strongloop/loopback-next/blob/docs/describe-how-to-test-tsc-tslint-changes/CONTRIBUTING.md
- https://github.com/strongloop/loopback-next/blob/docs/describe-how-to-test-tsc-tslint-changes/docs/DEVELOPING.md
- https://github.com/strongloop/loopback-next/blob/docs/describe-how-to-test-tsc-tslint-changes/docs/VSCODE.md

b-admike commented on docs/VSCODE.md:
```diff
@@ -0,0 +1,93 @@
+7. Open "Tasks" OUTPUT window and verify that compilation error was parsed by VSCode.
+
+8. Verify that compilation errors are correctly associated with the problematic source code line.
+   _(2018-02-82: this is does not work now, `tsc` is reporting paths relative to individual package directories.)_
```
typo if `2018-02-82` is a date here. I suggest we link it to an issue, but if it's fixed, we can get rid of this message.

bajtos replied:
Will fix. I opened https://github.com/strongloop/loopback-next/issues/1010 to keep track of this bug.

b-admike commented on docs/VSCODE.md:
```diff
@@ -0,0 +1,93 @@
+
+#### Refactoring in VS Code
+
+Verify that refactorings like "rename" will change all places using the renamed entity. Two different scenarios to verify: rename at the place where the entity is defined, rename at the place where the entity is used. (You can e.g. rename `inject` to test.)
```
I think for beginners saying `rename symbol` or F2 might be better than rename.

raymondfeng commented on docs/apidocs.html:
```diff
@@ -0,0 +1,33 @@
+      <h2>LoopBack4 packages (local)</h2>
+
+      <p>Run the following command to (re)build API doc files:
+      <p><pre><code>$ lerna run build:apidocs</code></pre>
```
We probably should use `npx lerna run build:apidocs` or add a script to top level package.json.

bajtos replied:
Good catch! I have `./node_modules/.bin` in my PATH and my Bash configured to run `npx` for commands not found, but I agree that's not something we can rely on in our docs.

raymondfeng commented:
FYI : https://github.com/strongloop/loopback-next/pull/1005. Feel free to reference it.

bajtos commented:
> FYI : #1005. Feel free to reference it.

I'll open a new pull request for that, I really want to get this on merged.
